OrbitVulkanLayer/DispatchTable.h

```diff
@@ -0,0 +1,413 @@
+    absl::ReaderMutexLock lock(&mutex_);
+    void* key = GetDispatchTableKey(dispatchable_object);
+    CHECK(device_dispatch_table_.contains(key));
+    CHECK(device_dispatch_table_.at(key).CmdBeginDebugUtilsLabelEXT != nullptr);
+    return device_dispatch_table_.at(key).CmdBeginDebugUtilsLabelEXT;
+  }
+
+  template <typename DispatchableType>
+  PFN_vkCmdEndDebugUtilsLabelEXT CmdEndDebugUtilsLabelEXT(
+      const DispatchableType& dispatchable_object) {
+    absl::ReaderMutexLock lock(&mutex_);
+    void* key = GetDispatchTableKey(dispatchable_object);
+    CHECK(device_dispatch_table_.contains(key));
+    CHECK(device_dispatch_table_.at(key).CmdEndDebugUtilsLabelEXT != nullptr);
+    return device_dispatch_table_.at(key).CmdEndDebugUtilsLabelEXT;
+  }
+
+  template <typename DispatchableType>
+  PFN_vkCmdDebugMarkerBeginEXT CmdDebugMarkerBeginEXT(const DispatchableType& dispatchable_object) {
+    absl::ReaderMutexLock lock(&mutex_);
+    void* key = GetDispatchTableKey(dispatchable_object);
+    CHECK(device_dispatch_table_.contains(key));
+    CHECK(device_dispatch_table_.at(key).CmdDebugMarkerBeginEXT != nullptr);
+    return device_dispatch_table_.at(key).CmdDebugMarkerBeginEXT;
+  }
+
+  template <typename DispatchableType>
+  PFN_vkCmdDebugMarkerEndEXT CmdDebugMarkerEndEXT(const DispatchableType& dispatchable_object) {
+    absl::ReaderMutexLock lock(&mutex_);
+    void* key = GetDispatchTableKey(dispatchable_object);
+    CHECK(device_dispatch_table_.contains(key));
+    CHECK(device_dispatch_table_.at(key).CmdDebugMarkerEndEXT != nullptr);
+    return device_dispatch_table_.at(key).CmdDebugMarkerEndEXT;
+  }
+
+  template <typename DispatchableType>
+  bool IsDebugMarkerExtensionSupported(const DispatchableType& dispatchable_object) {
+    absl::ReaderMutexLock lock(&mutex_);
+    void* key = GetDispatchTableKey(dispatchable_object);
+    CHECK(device_supports_debug_marker_extension_.contains(key));
+    return device_supports_debug_marker_extension_.at(key);
+  }
+
+  template <typename DispatchableType>
+  bool IsDebugUtilsExtensionSupported(const DispatchableType& dispatchable_object) {
+    absl::ReaderMutexLock lock(&mutex_);
+    void* key = GetDispatchTableKey(dispatchable_object);
+    CHECK(device_supports_debug_utils_extension_.contains(key));
+    return device_supports_debug_utils_extension_.at(key);
+  }
+
+ private:
+  // In vulkan, every "dispatchable type" has as a very first field in memory a pointer to the
+  // internal dispatch table. This pointer is unique per device/instance. So for example
+  // for a command buffer allocated on a certain device, this pointer is the same for the buffer
+  // and for the device. So we can use that pointer to uniquely map dispatchable types to their
+  // dispatch table.
+  template <typename DispatchableType>
+  void* GetDispatchTableKey(DispatchableType dispatchable_object) {
+    return *absl::bit_cast<void**>(dispatchable_object);
```

Comment:
This is a bit of a hack to access the first 8 bytes. Is there a chance that `hwvulkan_dispatch_t dispatch` of `struct VkInstance_T` and `struct VkDevice_T` (done a minor research here) is consistent enough that you can access it by name just with `dispatchable_object->dispatch`?

Reply:
Unfortunately that does not compile.

Reply:
I updated it using memcpy